### PR TITLE
[RFC] Refactor

### PR DESCRIPTION
**Pull request by iampersistent**

### CODE REVIEW

File: lib/Vespolina/Entity/Billing/BillingAgreement.php [hunk 1]
```diff
@@ -50,6 +102,44 @@ public function getBillingInterval()
...
+     */
+    public function setCreatedAt(\DateTime $createdAt)
+    {
+        $this->createdAt = $createdAt;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    function completeCurrentCycle(BillingRequestInterface $billingRequest)
```
Comment (inspiran) [inline lib/Vespolina/Entity/Billing/BillingAgreement.php:124]: We should move this to the billing manager and have it trigger an event (BILLING_CYCLE_COMPLETED).

File: lib/Vespolina/Entity/Billing/BillingAgreement.php [hunk 2]
```diff
@@ -62,6 +152,37 @@ public function getInitialBillingDate()
...
+    }
+
+    public function increaseNumberCyclesBilled()
+    {
+        $this->numberCyclesBilled++;
+
+        return $this;
+    }
+
     public function setOrder(OrderInterface $order)
```
Comment (inspiran) [inline lib/Vespolina/Entity/Billing/BillingAgreement.php:186]: Should be addOrder()  , a billing agreement could cover multiple orders.  (eg. if the customer creates two orders on the same day, you would want to use the same billing agreement.  You then would have a billing agreement with two billing request items, one pointing to the first order (item) and the second pointing to the newest order (item).

File: lib/Vespolina/Entity/Billing/BillingRequest.php [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    protected $paymentProfile;
+    protected $pricingSet;
+    protected $status;
+    protected $periodStart;
+    protected $periodEnd;
+
+    public function __construct()
+    {
+        $this->orderItems = array();
+        $this->status = self::STATUS_PENDING;
```
Comment (inspiran) [inline lib/Vespolina/Entity/Billing/BillingRequest.php:43]: It should have an initial (or draft) status.  Pending should be a status set when a sales clerk really agrees to offer it to a payment gateway.  Basically the process should be split in a) the generation of billing requests and b) fulfilling the billing request.  A sales clerk should be able to override / intercept billing requests after they have been created.  In some cases the clerk has to manually approve them before they become in state "pending".

File: lib/Vespolina/Entity/Invoice/Invoice.php [hunk 4]
```diff
@@ -28,6 +30,11 @@ class Invoice implements InvoiceInterface
     protected $periodEnd;
     protected $periodStart;
     protected $previousInvoice;
+    protected $paid = false;
```
Comment (inspiran) [inline lib/Vespolina/Entity/Invoice/Invoice.php:33]: what if the invoice has been partially paid?  We should have a status field here to indicate the payment status.

File: lib/Vespolina/Entity/Order/BaseOrder.php [hunk 5]
```diff
@@ -356,4 +373,24 @@ public function getUpdatedAt()
...
+     * @return boolean
+     */
+    public function isAllFree()
+    {
+        $items = $this->getItems();
+
+        foreach ($items as $i) {
+            /** @var Item $i */
+
+            if ($i->getProduct()->getName() != 'Free') {
```
Comment (inspiran) [inline lib/Vespolina/Entity/Order/BaseOrder.php:390]: Seems hackish, we should decouple orders from the pricing info. ... at least have a $product->isFree() method instead

File: lib/Vespolina/Entity/Billing/BillingAgreement.php (hunk 1 above)
Comment (iampersistent) [inline lib/Vespolina/Entity/Billing/BillingAgreement.php:124]: This is a self maintaining method, I'm not sure what the advantage is to moving it to the billing manager

File: lib/Vespolina/Entity/Partner/Partner.php [hunk 6]
```diff
@@ -24,11 +26,15 @@ class Partner implements PartnerInterface
     const ROLE_EMPLOYEE    = 'ROLE_EMPLOYEE';
     const ROLE_SUPPLIER    = 'ROLE_SUPPLIER';
 
+    const PAYMENT_PROFILE_TYPE_CREDIT_CARD = 'Credit Card';
+    const PAYMENT_PROFILE_TYPE_INVOICE = 'Invoice';
```
Comment (inspiran) [inline lib/Vespolina/Entity/Partner/Partner.php]: Invoice is not a payment profile.  it's just a legal document indicating that you need to pay (or have paid).

File: lib/Vespolina/Entity/Partner/PaymentProfile.php [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        return $this->billingAddress;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setBillingZipCode($billingZipCode)
+    {
+        $this->billingZipCode = $billingZipCode;
```
Comment (inspiran) [inline lib/Vespolina/Entity/Partner/PaymentProfile.php:97]: shouldn't this be $this->billingAddress->getZipcode() ?

File: lib/Vespolina/Form/Partner/AddressType.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Vespolina\Form\Partner;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+
+class AddressType extends AbstractType
```
Comment (inspiran) [inline lib/Vespolina/Form/Partner/AddressType.php:9]: we shouldn't have forms in the core.  That should go VespolinaPartnerBundle

File: lib/Vespolina/Entity/Billing/BillingAgreement.php (hunk 1 above)
Comment (inspiran) [inline lib/Vespolina/Entity/Billing/BillingAgreement.php:124]: To easily trigger the event.  But perhaps you are calling this method from the billing manager? Then it's ok.

**Pull request closed without merging**